builtin/providers/librato/resource_librato_metric.go

```diff
@@ -0,0 +1,432 @@
+	}
+	if a, ok := d.GetOk("description"); ok {
+		metric.Description = librato.String(a.(string))
+	}
+	if a, ok := d.GetOk("period"); ok {
+		metric.Period = librato.Uint(a.(uint))
+	}
+	if a, ok := d.GetOk("composite"); ok {
+		metric.Composite = librato.String(a.(string))
+	}
+
+	if a, ok := d.GetOk("attributes"); ok {
+
+		attributeData := a.([]interface{})
+		if len(attributeData) > 1 {
+			return fmt.Errorf("Only one set of attributes per metric is supported")
+		}
+
+		if len(attributeData) == 1 && attributeData[0] == nil {
+			return fmt.Errorf("No attributes found in attributes block")
+		}
+
+		attributeDataMap := attributeData[0].(map[string]interface{})
+		attributes := new(librato.MetricAttributes)
+
+		if v, ok := attributeDataMap["color"].(string); ok && v != "" {
+			attributes.Color = librato.String(v)
+		}
+		if v, ok := attributeDataMap["display_max"].(string); ok && v != "" {
+			attributes.DisplayMax = librato.String(v)
+		}
+		if v, ok := attributeDataMap["display_min"].(string); ok && v != "" {
+			attributes.DisplayMin = librato.String(v)
+		}
+		if v, ok := attributeDataMap["display_units_long"].(string); ok && v != "" {
+			attributes.DisplayUnitsLong = *librato.String(v)
+		}
+		if v, ok := attributeDataMap["display_units_short"].(string); ok && v != "" {
+			attributes.DisplayUnitsShort = *librato.String(v)
+		}
+		if v, ok := attributeDataMap["created_by_ua"].(string); ok && v != "" {
+			attributes.CreatedByUA = *librato.String(v)
+		}
+		if v, ok := attributeDataMap["display_stacked"].(bool); ok {
+			attributes.DisplayStacked = *librato.Bool(v)
+		}
+		if v, ok := attributeDataMap["gap_detection"].(bool); ok {
+			attributes.GapDetection = *librato.Bool(v)
+		}
+		if v, ok := attributeDataMap["aggregate"].(bool); ok {
+			attributes.Aggregate = *librato.Bool(v)
+		}
+
+		metric.Attributes = attributes
+	}
+
+	_, err := client.Metrics.Edit(metric)
+	if err != nil {
+		log.Printf("[INFO] ERROR creating Metric: %s", err)
+		return fmt.Errorf("Error creating Librato metric: %s", err)
+	}
+
+	resource.Retry(1*time.Minute, func() *resource.RetryError {
+		_, _, err := client.Metrics.Get(*metric.Name)
+		if err != nil {
+			if errResp, ok := err.(*librato.ErrorResponse); ok && errResp.Response.StatusCode == 404 {
+				return resource.RetryableError(err)
+			}
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	return metricReadResult(d, metric)
+}
+
+func resourceLibratoMetricRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*librato.Client)
+
+	id := d.Id()
+
+	log.Printf("[INFO] Reading Librato Metric: %s", id)
+	metric, _, err := client.Metrics.Get(id)
+	if err != nil {
+		if errResp, ok := err.(*librato.ErrorResponse); ok && errResp.Response.StatusCode == 404 {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error reading Librato Metric %s: %s", id, err)
+	}
+
+	log.Printf("[INFO] Read Librato Metric: %s", structToString(metric))
```

Comment:
If I use the `structToString` method I get this logged:
```
Librato fullMetric: {"name":"tftest-metric-vbljjgu6ys","description":"An updated test counter metric","type":"counter","attributes":{"color":null,"display_max":null,"display_min":null,"display_units_long":"","display_units_short":"","display_stacked":true,"display_transform":"","created_by_ua":"go-librato/0.1"}}
```
If I use `%v`, I get this logged:
```
2017/05/25 11:40:17 [INFO] Librato fullMetric: &{0xc4204ac520 0xc4204ad060 0xc4204ac530 <nil> <nil> <nil> 0xc42026ab80}
```
If I use `%#v`, I get
```
2017/05/25 11:42:12 [INFO] Librato fullMetric: &librato.Metric{Name:(*string)(0xc420481890), Description:(*string)(0xc42048e3a0), Type:(*string)(0xc4204818a0), Period:(*uint)(nil), DisplayName:(*string)(nil), Composite:(*string)(nil), Attributes:(*librato.MetricAttributes)(0xc420500480)}
```
If I use `%+v`, I get:
```
2017/05/25 11:43:15 [INFO] Librato fullMetric: &{Name:0xc4202f4020 Description:0xc4202f49c0 Type:0xc4202f4030 Period:<nil> DisplayName:<nil> Composite:<nil> Attributes:0xc420448000}
```
As you might expect, I prefer the first one. Nested structs are a PITA in terms of dumping/displaying them.